### PR TITLE
[FrameworkBundle] fixed custom domain for translations in php templates

### PR DESCRIPTION
**Pull request by robinlehrmann**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20135
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 1]
```diff
@@ -37,7 +39,29 @@ class PhpExtractor extends AbstractFileExtractor implements ExtractorInterface
      *
      * @var array
      */
-    protected $sequences = array(
+    protected static $sequences = array(
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: The `static` change should be reverted.

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 2]
```diff
@@ -153,26 +198,38 @@ protected function parseTokens($tokens, MessageCatalogue $catalog)
         $tokenIterator = new \ArrayIterator($tokens);
 
         for ($key = 0; $key < $tokenIterator->count(); ++$key) {
-            foreach ($this->sequences as $sequence) {
+            foreach (self::$sequences as $sequence) {
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: Must then be reverted too.

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 3]
```diff
@@ -105,13 +129,34 @@ private function seekToNextRelevantToken(\Iterator $tokenIterator)
         }
     }
 
+    private function seekToMessageParamsEnd(\Iterator $tokenIterator)
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: maybe better name this `seekBehindMessageParams`

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 4]
```diff
@@ -85,7 +109,7 @@ public function setPrefix($prefix)
      */
     protected function normalizeToken($token)
     {
-        if (isset($token[1]) && 'b"' !== $token) {
+        if ('b"' !== $token && isset($token[1])) {
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: IMO this should be reverted as it is not part of the bug fix.

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 5]
```diff
@@ -105,13 +129,34 @@ private function seekToNextRelevantToken(\Iterator $tokenIterator)
...
+    }
+
     /**
-     * Extracts the message from the iterator while the tokens
-     * match allowed message tokens.
+     * Extracts the raw string from the iterator while the tokens
+     * match allowed tokens.
      */
-    private function getMessage(\Iterator $tokenIterator)
+    private function provideRawString(\Iterator $tokenIterator)
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: What about `getStringValue()` instead?

**Review by xabbuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 6]
```diff
@@ -105,13 +129,34 @@ private function seekToNextRelevantToken(\Iterator $tokenIterator)
...
     }
 
+    private function seekToMessageParamsEnd(\Iterator $tokenIterator)
+    {
+        $openBraces = 0;
+
+        for (; $tokenIterator->valid(); $tokenIterator->next()) {
+            $t = $tokenIterator->current();
+
+            if ('[' === $t[0] || '(' === $t[0]) {
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: you should not take account of braces inside a string
```php
->trans('msg', [ 'p' => '[ \'' ], 'not_messages')
```

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 7]
```diff
@@ -37,7 +39,29 @@ class PhpExtractor extends AbstractFileExtractor implements ExtractorInterface
...
+            ',',
+            self::DOMAIN_TOKEN,
+        ),
+        array(
+            '->',
+            'transChoice',
+            '(',
+            self::MESSAGE_TOKEN,
+            ',',
+            T_LNUMBER,
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: it can be an expression or function call

**Review by aitboudad — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 7 above)
Comment (robinlehrmann) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: I added a new test with array_map function and the tests passes. Can you explain what you mean please :) ?

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 6 above)
Comment (robinlehrmann) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: I added this line to my tests too and it passes.

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 4 above)
Comment (robinlehrmann) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: Done

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 3 above)
Comment (robinlehrmann) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: Done

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 5 above)
Comment (robinlehrmann) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: Done

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 2 above)
Comment (robinlehrmann) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: Done

**Review by robinlehrmann — COMMENTED** [review]
I added some tests and it passes.

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 7 above)
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]:
```php
<?php echo $view['translator']->transChoice('msg', 10 + 1, [], 'not_messages'); ?>
<?php echo $view['translator']->transChoice('msg', intval(4.5), [], 'not_messages'); ?>
```

Comment (robinlehrmann) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: Okay, I added this lines and the Tests don't pass. Thank you!
I will fix it asap.

Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: I guess you can use the same approach here that is already used for skipping the message params.

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 6 above)
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: @aitboudad In your example, `$t` will be an array consisting of three elements where the value at index 0 is `T_CONSTANT_ENCAPSED_STRING` if I am not mistaken.

**Comment by robinlehrmann** [comment]
Hi, @aitboudad @xabbuh 
I've fixed the issue and added some tests for it. Could you review my changes please ?

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 8]
```diff
@@ -24,6 +24,8 @@
 class PhpExtractor extends AbstractFileExtractor implements ExtractorInterface
 {
     const MESSAGE_TOKEN = 300;
+    const MESSAGE_ARGUMENTS_TOKEN = 1000;
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: I would name this `METHOD_ARGUMENTS_TOKEN`

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: I would now rename this too (something like `skipMethodArgument()`)

**Review by xabbuh — APPROVED** [review]
:+1:

Status: Reviewed

**Review by aitboudad — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @robinlehrmann.

**Pull request merged**